### PR TITLE
[#3791] one image upload

### PR DESCRIPTION
**Pull request by valllllll2000**
#### The solution
Upload one image to s3 and insert the answer to the store
Added tests
#### Reviewer Checklist
* [ ] Added an explanation about the work done
* [ ] Connected the PR and the issue on Zenhub
* [ ] Added a test plan to the issue
* [ ] Updated the copyright header (when relevant)
* [ ] Formatted the code
* [ ] Added a documentation (if relevant)
* [ ] Added some unit tests (if relevant)

### CODE REVIEW

File: GAE/src/org/akvo/flow/rest/ImageUploadRestService.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            return null;
+        }
+    }
+
+    private void saveQuestionAnswer(Question question, SurveyInstance formInstance, String fileName) {
+        Long questionGroupId = question.getQuestionGroupId();
+        QuestionGroup group = new QuestionGroupDao().getByKey(questionGroupId);
+        boolean repeatable = group.getRepeatable();
+        if (repeatable) {
+            saveRepeatedQuestionAnswer(question, formInstance, fileName);
```
Comment (muloem) [inline GAE/src/org/akvo/flow/rest/ImageUploadRestService.java:128]: I think we should leave out the handling of repeat question group for now.  In this case, there are no photos in the repeated groups I believe. and its one that needs a little more designing on how to handle it.

File: GAE/src/org/akvo/flow/rest/ImageUploadRestService.java [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        createAndSaveQuestionAnswer(question, formInstance, fileName, questionAnswerStoreDao, iteration);
+    }
+
+    private void createAndSaveQuestionAnswer(Question question, SurveyInstance formInstance, String fileName, QuestionAnswerStoreDao questionAnswerStoreDao, int iteration) {
+        QuestionAnswerStore store = new QuestionAnswerStore();
+        store.setSurveyId(formInstance.getSurveyId());
+        store.setSurveyInstanceId(formInstance.getKey().getId());
+        store.setQuestionID(question.getKey().getId() + "");
+
+        //TODO: should the date be same as collection date on the form Instance??
```
Comment (muloem) [inline GAE/src/org/akvo/flow/rest/ImageUploadRestService.java:165]: If we are able to get the metadata of when the image was taken maybe thats better.

**Review by muloem — COMMENTED** [review]
Looks good! 👍 

A couple of comments inline to consider and we also need to handle the extraction and saving of the geo data if its there in the image but we can maybe do that in a follow up PR

File: GAE/src/org/akvo/flow/rest/ImageUploadRestService.java (hunk 2 above)
Comment (valllllll2000) [reply GAE/src/org/akvo/flow/rest/ImageUploadRestService.java:132]: I will add this together with geotag extraction in second PR

File: GAE/src/org/akvo/flow/rest/ImageUploadRestService.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+@Controller
+@RequestMapping("/image_upload")
+public class ImageUploadRestService {
+
+    private static final Logger log = Logger
+            .getLogger(ImageUploadRestService.class.getName());
+
+    @RequestMapping(method = RequestMethod.POST, value = "/question/{questionId}/instance/{instanceId}")
+    @ResponseBody
+    public Response uploadImage(@PathVariable("questionId") String questionId,
```
Comment (muloem) [inline GAE/src/org/akvo/flow/rest/ImageUploadRestService.java:49]: I believe its possible to already convert to a `Long` here.. Same for the `formInstanceId`

Comment (muloem) [reply GAE/src/org/akvo/flow/rest/ImageUploadRestService.java:49]:
```suggestion
    public Response uploadImage(@PathVariable("questionId") Long questionId,
```

File: GAE/src/org/akvo/flow/rest/ImageUploadRestService.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        }
+
+        if (file == null || file.isEmpty()) {
+            return new Response(400, "File is not valid");
+        }
+        String fileExtension = getFileType(file);
+        if (fileExtension == null) {
+            return new Response(400, "File type is not valid: only jpg and png are accepted");
+        }
+        //TODO: shall we resize? does the calling task handle resize
```
[the remark this answers is not in the document]
Comment (valllllll2000) [reply GAE/src/org/akvo/flow/rest/ImageUploadRestService.java:78]: https://github.com/akvo/akvo-flow/issues/3817

File: GAE/src/org/akvo/flow/rest/ImageUploadRestService.java (hunk 3 above)
Comment (valllllll2000) [reply GAE/src/org/akvo/flow/rest/ImageUploadRestService.java]: done

**Review by muloem — APPROVED** [review]

**Pull request merged**